go: Apply the options and correct the bitmasks

qinqon opened the pull request:
The golang binding was not applying the option that were passed and also the bitmask was not correct since it has to start from "2". It also create a pair of apply test for the golang binding and activate the logs at the clib.

Depends on:
- https://github.com/nmstate/nmstate/pull/1811

qinqon committed:
go: Add apply check

To fully cover golang binding functionality is needed to test the apply
nmstate primitive. This change add that check creating a dummy
interface.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon committed:
go: Apply options to nmstate

The options were being passed but not applied also the bitmask constant
need to start from "2".

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon committed:
go: Add apply with commit test

Signed-off-by: Quique Llorente <ellorent@redhat.com>

cathay4t commented on rust/src/clib/lib.rs:
```diff
@@ -167,7 +170,7 @@ pub extern "C" fn nmstate_net_state_apply(
     net_state.set_timeout(rollback_timeout);
 
-    // TODO: save log to the output pointer
+    let _ = env_logger::try_init();
```
This just pring logs to stderr which should not be done by a C library.
As the API designed, the log should saved to `char *`.

Initial plan is mimic what was done by https://github.com/gahag/memory_logger/

qinqon replied:
Ahh that's the logger I was looking for, thanks! I will try to integrate it.

qinqon replied:
@cathay4t Looks like the tokio-rs projects has a tracing library with non-global tracer https://github.com/tokio-rs/tracing, you can define the scope of a trace and run code there.

Maybe  nmstate rust can use that library don't know if we can make it use memory or retrieve what has being logged.

cathay4t replied:
The tokio tracing seems overkill for this project. Please try memory logger. But this can be done in another PR. For this PR, just remove the print in stdout.

cathay4t commented on rust/src/clib/nmstate.h.in:
```diff
@@ -186,92 +186,20 @@ int nmstate_checkpoint_rollback(const char *checkpoint, char **log, char **err_k
-#ifdef __cplusplus
-} /* extern "C" */
-#endif
+void nmstate_cstring_free(char *cstring);
```
This is unrelated to this PR.
Unless good reasoning, I don't like to the C API/ABI.
Current functions sames good to me.

qinqon replied:
This PR depends on https://github.com/nmstate/nmstate/pull/1795, I can put common code under a new functions and convert the current functions as wrappers that call it.

cathay4t commented on rust/src/go/nmstate/Makefile:
```diff
@@ -1,13 +1,18 @@
-	go test;
-
+	DESTDIR=$(TMPDIR) make -C $(TOP_SRC_DIR) install
+	sudo RUST_LOG=info LD_LIBRARY_PATH=$(TMPDIR)/usr/local/lib64 CGO_CFLAGS="-I$(TMPDIR)/usr/local/include" CGO_LDFLAGS="-L$(TMPDIR)/usr/local/lib64" go test $(WHAT);
```
Try to make text width 78.

cathay4t reviewed (changes requested):
The go code is reviewed as I am not familiar it yet.

cathay4t commented:
Please enable CI test for go binding since you included the test case.

qinqon committed:
rust,go: Run make check on ci

The clib is being implicitly tested running the python tests, this
change also add the golang binding as part of the main test matrix to
test agains main and stable NM.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon commented:
> Please enable CI test for go binding since you included the test case.

Calling `make -C rust check` at last commit.

cathay4t commented on rust/src/clib/test/Makefile:
```diff
@@ -1,36 +1,42 @@
 	ln -sfv $(CLIB_SO_FULL) $(TMPDIR)/$(CLIB_SO_DEV)
 
-nmstate_test: nmstate_test.c nmstate.h libnmstate.so
+nmstate_test: nmstate_test.c nmstate.h install
```
Please use `CGO_LDFLAGS` and `CGO_CFLAGS` instead of install it. I ran `make check` in my OS which I don't want it been installed in this case

qinqon replied:
I will copy (although it introduces some redundancy) the build/install code the the src/go/nmstate/Makefile
